### ISSUE TITLE
Using crates.io libc triggers unstable libc warning

### ISSUE THREAD

**Issue by SSheldon** [rust-lang/rust#21771]
With the crates.io distribution of libc (#20606) in my Cargo.toml:
```
[dependencies]
libc = "0.1"
```

I still get a warning on `extern crate libc`:

```
lib.rs:6:1: 6:19 warning: use of unstable library feature 'libc'
lib.rs:6 extern crate libc;
         ^~~~~~~~~~~~~~~~~~
lib.rs:6:1: 6:19 help: add #![feature(libc)] to the crate attributes to silence this warning
lib.rs:6 extern crate libc;
         ^~~~~~~~~~~~~~~~~~
```

Am I linking to the wrong crate? How do you choose the version from cargo over the version included in rustc?

**Comment by nagisa** [comment]
A rule of thumb is to always use the version shipped with the compiler unless you know you need the version from crates.io (e.g. crossbuilding).

The warning is expected since libc in crate is the same thing that is shipped with the rust compiler.

**Comment by SSheldon** [comment]
@nagisa, #19260 says of libc:

> These crates will all be distributed, but participate in #![staged_api]. This means they will not be accessible in the stable channel's standard distribution, but they will all be available through crates.io.

So that suggests to me that the version shipped with the compiler shouldn't be used.

**Comment by SSheldon** [comment]
cc @alexcrichton, who has worked on the libc for crates.io

**Comment by alexcrichton** [comment]
Working on a fix, thanks for the report! I've also seen this recently and I've been baffled.

**Comment by tbu-** [comment]
@nagisa You shouldn't work with the explicitely-deprecated compiler-shipped libraries.

**Comment by stevedonovan** [comment]
I note that this is still the case for 1.12 stable, which seems alarming...

**Comment by steveklabnik** [comment]
@stevedonovan can you share your exact code? I can't reproduce.

**Comment by stevedonovan** [comment]
Sorry, I forgot to actually add libc to my crate that imported a crate needing libc, hence the issue! That makes sense, so I'm happy.

**Comment by steveklabnik** [comment]
Ah, great!
